File: src/clients/desktop_app/ami_daemon/base_agent/memory/session_manager.py
"""
Session Manager for Conversation Persistence

Simple session-based conversation management:
- Each session is a JSONL file
- Sessions timeout after 30 minutes of inactivity
- New sessions load context from previous session

Inspired by OpenClaw's session management pattern.
"""

import json
import logging
import os
import uuid
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)

# Session timeout in minutes
SESSION_TIMEOUT_MINUTES = 30

# Number of messages to carry over to new session
CONTEXT_MESSAGES_COUNT = 5


def generate_session_id() -> str:
    """Generate a unique session ID."""
    return f"sess_{uuid.uuid4().hex[:12]}"


def _utc_now_iso() -> str:
    """Get current UTC time as ISO string."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def generate_message_id() -> str:
    """Generate a unique message ID."""
    return f"msg_{uuid.uuid4().hex[:8]}"


class SessionManager:
    """
    Manages conversation sessions with automatic timeout handling.

    Usage:
        manager = SessionManager()

        # Get or create session (handles timeout automatically)
        session = manager.get_active_session()

        # Append message
        manager.append_message("user", "Hello")
        manager.append_message("assistant", "Hi there!")

        # Load messages for display
        messages = manager.get_recent_messages(limit=50)
    """

    def __init__(self, base_path: Optional[Path] = None):
        """
        Initialize session manager.

        Args:
            base_path: Base directory for sessions. Defaults to ~/.ami/sessions/
        """
        if base_path is None:
            base_path = Path.home() / ".ami" / "sessions"

        self.base_path = Path(base_path)
        self.base_path.mkdir(parents=True, exist_ok=True)

        self.index_path = self.base_path / "index.json"
        self._index_cache: Optional[Dict] = None

    # ==================== Index Management ====================

    def _load_index(self) -> Dict:
        """Load session index from disk."""
        if self._index_cache is not None:
            return self._index_cache

        if not self.index_path.exists():
            self._index_cache = {
                "current_session_id": None,
                "last_activity": None,
                "sessions": {},
            }
            return self._index_cache

        try:
            with open(self.index_path, "r", encoding="utf-8") as f:
                self._index_cache = json.load(f)
                return self._index_cache
        except (json.JSONDecodeError, IOError) as e:
            logger.error(f"Failed to load session index: {e}")
            self._index_cache = {
                "current_session_id": None,
                "last_activity": None,
                "sessions": {},
            }
            return self._index_cache

    def _save_index(self) -> None:
        """Save session index to disk."""
        if self._index_cache is None:
            return

        try:
            with open(self.index_path, "w", encoding="utf-8") as f:
                json.dump(self._index_cache, f, indent=2, ensure_ascii=False)
        except IOError as e:
            logger.error(f"Failed to save session index: {e}")

    def _invalidate_cache(self) -> None:
        """Invalidate index cache."""
        self._index_cache = None

    # ==================== Session Lifecycle ====================

    def _is_session_expired(self) -> bool:
        """Check if current session has expired (30 min timeout)."""
        index = self._load_index()

        last_activity_str = index.get("last_activity")
        if not last_activity_str:
            logger.debug("Session expired: no last_activity")
            return True

        try:
            # Parse ISO format timestamp
            if last_activity_str.endswith("Z"):
                last_activity_str = last_activity_str[:-1] + "+00:00"
            last_activity = datetime.fromisoformat(last_activity_str)

            # Ensure timezone aware comparison
            now = datetime.now(timezone.utc)
            if last_activity.tzinfo is None:
                last_activity = last_activity.replace(tzinfo=timezone.utc)

            timeout = timedelta(minutes=SESSION_TIMEOUT_MINUTES)
            elapsed = now - last_activity
            expired = elapsed > timeout

            if expired:
                logger.debug(f"Session expired: last_activity={last_activity}, elapsed={elapsed}")

            return expired
        except (ValueError, TypeError) as e:
            logger.warning(f"Failed to parse last_activity: {e}")
            return True

    def _create_new_session(self, carry_context: bool = True) -> str:
        """
        Create a new session.

        Args:
            carry_context: Whether to carry messages from previous session

        Returns:
            New session ID
        """
        index = self._load_index()
        previous_session_id = index.get("current_session_id")

        # Generate new session
        new_session_id = generate_session_id()
        now = _utc_now_iso()

        # Create session file with header
        session_file = self.base_path / f"{new_session_id}.jsonl"
        header = {
            "type": "header",
            "session_id": new_session_id,
            "created_at": now,
            "previous_session_id": previous_session_id,
        }

        with open(session_file, "w", encoding="utf-8") as f:
            f.write(json.dumps(header, ensure_ascii=False) + "\n")

        # Update index
        index["current_session_id"] = new_session_id
        index["last_activity"] = now
        index["sessions"][new_session_id] = {
            "created_at": now,
            "updated_at": now,
            "message_count": 0,
            "previous_session_id": previous_session_id,
        }
        self._save_index()

        logger.info(f"Created new session: {new_session_id}")

        # Carry context from previous session
        if carry_context and previous_session_id:
            self._carry_context_from_previous(previous_session_id, new_session_id)

        return new_session_id

    def _carry_context_from_previous(self, previous_id: str, new_id: str) -> None:
        """Copy last N messages from previous session to new session as context."""
        previous_file = self.base_path / f"{previous_id}.jsonl"
        if not previous_file.exists():
            return

        # Read last N messages from previous session
        messages = []
        try:
            with open(previous_file, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        record = json.loads(line)
                        if record.get("type") == "message":
                            messages.append(record)
                    except json.JSONDecodeError:
                        continue
        except IOError:
            return

        # Get last N messages
        context_messages = messages[-CONTEXT_MESSAGES_COUNT:]
        if not context_messages:
            return

        # Append to new session as context
        new_file = self.base_path / f"{new_id}.jsonl"
        with open(new_file, "a", encoding="utf-8") as f:
            for msg in context_messages:
                # Mark as context from previous session
                msg["is_context"] = True
                msg["from_session"] = previous_id
                f.write(json.dumps(msg, ensure_ascii=False) + "\n")

        logger.debug(f"Carried {len(context_messages)} messages from {previous_id} to {new_id}")

    def get_active_session(self) -> str:
        """
        Get or create active session.

        If current session is expired (> 30 min), creates new session
        and carries context from previous session.

        Returns:
            Active session ID
        """
        index = self._load_index()

        # No current session or expired
        if not index.get("current_session_id") or self._is_session_expired():
            return self._create_new_session(carry_context=True)

        return index["current_session_id"]

    def force_new_session(self) -> str:
        """
        Force create a new session regardless of timeout.

        Returns:
            New session ID
        """
        return self._create_new_session(carry_context=True)

    # ==================== Message Operations ====================

    def append_message(
        self,
        role: str,
        content: str,
        message_id: Optional[str] = None,
        attachments: Optional[List[Dict]] = None,
        metadata: Optional[Dict] = None,
    ) -> str:
        """
        Append a message to the current session.

        Args:
            role: Message role (user, assistant, system)
            content: Message content
            message_id: Optional message ID (auto-generated if not provided)
            attachments: Optional attachments
            metadata: Optional metadata

        Returns:
            Message ID
        """
        session_id = self.get_active_session()
        session_file = self.base_path / f"{session_id}.jsonl"

        if message_id is None:
            message_id = generate_message_id()

        now = _utc_now_iso()

        message = {
            "type": "message",
            "id": message_id,
            "role": role,
            "content": content,
            "timestamp": now,
        }

        if attachments:
            message["attachments"] = attachments
        if metadata:
            message["metadata"] = metadata

        # Append to file
        with open(session_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(message, ensure_ascii=False) + "\n")

        # Update index
        index = self._load_index()
        index["last_activity"] = now
        if session_id in index["sessions"]:
            index["sessions"][session_id]["updated_at"] = now
            index["sessions"][session_id]["message_count"] = \
                index["sessions"][session_id].get("message_count", 0) + 1
        self._save_index()

        return message_id

    def get_messages(self, session_id: Optional[str] = None, limit: int = 100) -> List[Dict]:
        """
        Get messages from a session.

        Args:
            session_id: Session ID (defaults to current session)
            limit: Maximum messages to return

        Returns:
            List of message dicts (oldest to newest, limited to most recent)
        """
        if session_id is None:
            index = self._load_index()
            session_id = index.get("current_session_id")
            if not session_id:
                return []

        session_file = self.base_path / f"{session_id}.jsonl"
        if not session_file.exists():
            return []

        messages = []
        try:
            with open(session_file, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        record = json.loads(line)
                        if record.get("type") == "message":
                            messages.append(record)
                    except json.JSONDecodeError:
                        continue
        except IOError:
            return []

        # Return most recent messages
        if len(messages) > limit:
            return messages[-limit:]
        return messages

    def get_recent_messages(self, limit: int = 50) -> List[Dict]:
        """
        Get recent messages from current session.

        Convenience method for loading messages on app start.

        Args:
            limit: Maximum messages to return

        Returns:
            List of recent messages
        """
        return self.get_messages(limit=limit)

    # ==================== Session Info ====================

    def get_current_session_id(self) -> Optional[str]:
        """Get current session ID without triggering timeout check."""
        index = self._load_index()
        return index.get("current_session_id")

    def get_session_info(self, session_id: Optional[str] = None) -> Optional[Dict]:
        """Get session metadata."""
        if session_id is None:
            session_id = self.get_current_session_id()
        if not session_id:
            return None

        index = self._load_index()
        return index.get("sessions", {}).get(session_id)

    def list_sessions(self, limit: int = 20) -> List[Dict]:
        """
        List recent sessions.

        Args:
            limit: Maximum sessions to return

        Returns:
            List of session info dicts, sorted by updated_at descending
        """
        index = self._load_index()
        sessions = []

        for session_id, info in index.get("sessions", {}).items():
            sessions.append({
                "session_id": session_id,
                **info,
            })

        # Sort by updated_at descending
        sessions.sort(key=lambda s: s.get("updated_at", ""), reverse=True)

        return sessions[:limit]

    def touch_session(self) -> None:
        """Update last_activity timestamp without adding a message."""
        index = self._load_index()
        if index.get("current_session_id"):
            now = _utc_now_iso()
            index["last_activity"] = now
            self._save_index()

File: src/clients/desktop_app/daemon-ts/src/services/session-manager.ts
/**
 * Session Manager — JSONL-based conversation persistence.
 *
 * Ported from session_manager.py.
 *
 * - Each session is a JSONL file in ~/.ami/sessions/
 * - Sessions timeout after 30 minutes of inactivity
 * - New sessions carry context (last 5 messages) from previous session
 * - Cursor-based cross-session history traversal
 */

import {
  existsSync,
  mkdirSync,
  readFileSync,
  writeFileSync,
  appendFileSync,
  unlinkSync,
  readdirSync,
} from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import { randomUUID } from "node:crypto";
import { createLogger } from "../utils/logging.js";

const logger = createLogger("session-manager");

// ===== Constants =====

const SESSION_TIMEOUT_MINUTES = 30;
const CONTEXT_MESSAGES_COUNT = 5;
const SESSIONS_DIR = join(homedir(), ".ami", "sessions");

// ===== Helpers =====

function generateSessionId(): string {
  return `sess_${randomUUID().replace(/-/g, "").slice(0, 12)}`;
}

function generateMessageId(): string {
  return `msg_${randomUUID().replace(/-/g, "").slice(0, 8)}`;
}

function utcNowISO(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function sanitizeSessionId(id: string): string {
  return id.replace(/[/\\]/g, "_").replace(/\.\./g, "_");
}

// ===== Types =====

export interface SessionMessage {
  type: "message";
  id: string;
  role: string;
  content: string;
  timestamp: string;
  attachments?: Record<string, unknown>[];
  metadata?: Record<string, unknown>;
  is_context?: boolean;
  from_session?: string;
}

interface SessionHeader {
  type: "header";
  session_id: string;
  created_at: string;
  previous_session_id: string | null;
}

interface SessionInfo {
  created_at: string;
  updated_at: string;
  message_count: number;
  previous_session_id: string | null;
}

interface SessionIndex {
  current_session_id: string | null;
  last_activity: string | null;
  sessions: Record<string, SessionInfo>;
}

// ===== SessionManager =====

export class SessionManager {
  readonly basePath: string;
  private indexPath: string;
  private indexCache: SessionIndex | null = null;

  constructor(basePath?: string) {
    this.basePath = basePath ?? SESSIONS_DIR;
    mkdirSync(this.basePath, { recursive: true });
    this.indexPath = join(this.basePath, "index.json");
  }

  // ===== Index Management =====

  private loadIndex(): SessionIndex {
    if (this.indexCache) return this.indexCache;

    if (!existsSync(this.indexPath)) {
      this.indexCache = {
        current_session_id: null,
        last_activity: null,
        sessions: {},
      };
      return this.indexCache;
    }

    try {
      const raw = readFileSync(this.indexPath, "utf-8");
      this.indexCache = JSON.parse(raw) as SessionIndex;
      return this.indexCache;
    } catch {
      this.indexCache = {
        current_session_id: null,
        last_activity: null,
        sessions: {},
      };
      return this.indexCache;
    }
  }

  private saveIndex(): void {
    if (!this.indexCache) return;
    try {
      writeFileSync(
        this.indexPath,
        JSON.stringify(this.indexCache, null, 2),
        "utf-8",
      );
    } catch (err) {
      logger.error({ err }, "Failed to save session index");
    }
  }

  // ===== Session Lifecycle =====

  isSessionExpired(): boolean {
    const index = this.loadIndex();
    const lastActivity = index.last_activity;
    if (!lastActivity) return true;

    try {
      const lastMs = new Date(lastActivity).getTime();
      const nowMs = Date.now();
      return nowMs - lastMs > SESSION_TIMEOUT_MINUTES * 60_000;
    } catch {
      return true;
    }
  }

  private createNewSession(carryContext: boolean = true): string {
    const index = this.loadIndex();
    const previousSessionId = index.current_session_id;

    const newSessionId = generateSessionId();
    const now = utcNowISO();

    // Create session JSONL file with header (sessionId is already safe — generated internally)
    const sessionFile = join(this.basePath, `${sanitizeSessionId(newSessionId)}.jsonl`);
    const header: SessionHeader = {
      type: "header",
      session_id: newSessionId,
      created_at: now,
      previous_session_id: previousSessionId,
    };
    writeFileSync(sessionFile, JSON.stringify(header) + "\n", "utf-8");

    // Update index
    index.current_session_id = newSessionId;
    index.last_activity = now;
    index.sessions[newSessionId] = {
      created_at: now,
      updated_at: now,
      message_count: 0,
      previous_session_id: previousSessionId,
    };
    this.saveIndex();

    logger.info({ sessionId: newSessionId }, "Created new session");

    // Carry context from previous session
    if (carryContext && previousSessionId) {
      this.carryContextFromPrevious(previousSessionId, newSessionId);
    }

    return newSessionId;
  }

  private carryContextFromPrevious(
    previousId: string,
    newId: string,
  ): void {
    const previousFile = join(this.basePath, `${sanitizeSessionId(previousId)}.jsonl`);
    if (!existsSync(previousFile)) return;

    const messages = this.readSessionMessages(previousId);
    const contextMessages = messages.slice(-CONTEXT_MESSAGES_COUNT);
    if (contextMessages.length === 0) return;

    const newFile = join(this.basePath, `${sanitizeSessionId(newId)}.jsonl`);
    for (const msg of contextMessages) {
      msg.is_context = true;
      msg.from_session = previousId;
      appendFileSync(newFile, JSON.stringify(msg) + "\n", "utf-8");
    }

    logger.debug(
      { count: contextMessages.length, from: previousId, to: newId },
      "Carried context messages",
    );
  }

  getActiveSession(): string {
    const index = this.loadIndex();
    if (!index.current_session_id || this.isSessionExpired()) {
      return this.createNewSession(true);
    }
    return index.current_session_id;
  }

  forceNewSession(): string {
    return this.createNewSession(true);
  }

  getCurrentSessionId(): string | null {
    const index = this.loadIndex();
    return index.current_session_id;
  }

  // ===== Message Operations =====

  appendMessage(
    role: string,
    content: string,
    messageId?: string,
    attachments?: Record<string, unknown>[],
    metadata?: Record<string, unknown>,
  ): string {
    const sessionId = this.getActiveSession();
    const sessionFile = join(this.basePath, `${sanitizeSessionId(sessionId)}.jsonl`);

    const id = messageId ?? generateMessageId();
    const now = utcNowISO();

    const message: SessionMessage = {
      type: "message",
      id,
      role,
      content,
      timestamp: now,
    };

    if (attachments) message.attachments = attachments;
    if (metadata) message.metadata = metadata;

    appendFileSync(sessionFile, JSON.stringify(message) + "\n", "utf-8");

    // Update index
    const index = this.loadIndex();
    index.last_activity = now;
    const info = index.sessions[sessionId];
    if (info) {
      info.updated_at = now;
      info.message_count = (info.message_count || 0) + 1;
    }
    this.saveIndex();

    return id;
  }

  getMessages(sessionId?: string, limit: number = 100): SessionMessage[] {
    if (!sessionId) {
      const index = this.loadIndex();
      sessionId = index.current_session_id ?? undefined;
      if (!sessionId) return [];
    }

    const messages = this.readSessionMessages(sessionId);
    if (messages.length > limit) {
      return messages.slice(-limit);
    }
    return messages;
  }

  // ===== Session Info =====

  getSessionInfo(
    sessionId?: string,
  ): SessionInfo | null {
    if (!sessionId) {
      sessionId = this.getCurrentSessionId() ?? undefined;
    }
    if (!sessionId) return null;

    const index = this.loadIndex();
    return index.sessions[sessionId] ?? null;
  }

  listSessions(limit: number = 20): Array<SessionInfo & { session_id: string }> {
    const index = this.loadIndex();
    const sessions: Array<SessionInfo & { session_id: string }> = [];

    for (const [id, info] of Object.entries(index.sessions)) {
      sessions.push({ session_id: id, ...info });
    }

    sessions.sort((a, b) => b.updated_at.localeCompare(a.updated_at));
    return sessions.slice(0, limit);
  }

  touchSession(): void {
    const index = this.loadIndex();
    if (index.current_session_id) {
      const now = utcNowISO();
      index.last_activity = now;
      this.saveIndex();
    }
  }

  // ===== History Traversal =====

  getHistoryMessages(
    beforeTimestamp: string,
    limit: number = 30,
  ): { messages: SessionMessage[]; has_more: boolean; oldest_timestamp: string | null } {
    const index = this.loadIndex();
    const currentSessionId = index.current_session_id;
    if (!currentSessionId) {
      return { messages: [], has_more: false, oldest_timestamp: null };
    }

    let cursorMs: number;
    try {
      cursorMs = new Date(beforeTimestamp).getTime();
      if (isNaN(cursorMs)) {
        return { messages: [], has_more: false, oldest_timestamp: null };
      }
    } catch {
      return { messages: [], has_more: false, oldest_timestamp: null };
    }

    const collected: SessionMessage[] = [];
    let sessionId: string | null = currentSessionId;
    const visited = new Set<string>();
    let hasMoreSessions = false;

    while (sessionId) {
      if (visited.has(sessionId)) break;
      visited.add(sessionId);

      const messages = this.readSessionMessages(sessionId);

      for (const msg of messages) {
        // Skip context messages (duplicates from carry-over)
        if (msg.is_context) continue;
        if (!msg.timestamp) continue;

        const msgMs = new Date(msg.timestamp).getTime();
        if (isNaN(msgMs)) continue;

        if (msgMs < cursorMs) {
          collected.push(msg);
        }
      }

      // Move to previous session
      sessionId = this.getPreviousSessionId(sessionId);

      if (collected.length > limit && sessionId) {
        hasMoreSessions = true;
        break;
      }
    }

    // Sort by timestamp ascending, take most recent `limit`
    collected.sort((a, b) => a.timestamp.localeCompare(b.timestamp));

    const hasMore = collected.length > limit || hasMoreSessions;
    const resultMessages =
      collected.length > limit ? collected.slice(-limit) : collected;

    const oldestTimestamp =
      resultMessages.length > 0 ? resultMessages[0].timestamp : null;

    return { messages: resultMessages, has_more: hasMore, oldest_timestamp: oldestTimestamp };
  }

  // ===== Internal Helpers =====

  private readSessionMessages(sessionId: string): SessionMessage[] {
    const sessionFile = join(this.basePath, `${sanitizeSessionId(sessionId)}.jsonl`);
    if (!existsSync(sessionFile)) return [];

    const messages: SessionMessage[] = [];
    try {
      const content = readFileSync(sessionFile, "utf-8");
      for (const line of content.split("\n")) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        try {
          const record = JSON.parse(trimmed);
          if (record.type === "message") {
            messages.push(record as SessionMessage);
          }
        } catch {
          // Skip invalid lines
        }
      }
    } catch {
      // File read error
    }
    return messages;
  }

  private getPreviousSessionId(sessionId: string): string | null {
    const index = this.loadIndex();
    const info = index.sessions[sessionId];
    if (info?.previous_session_id) {
      return info.previous_session_id;
    }

    // Fallback: read from session file header
    const sessionFile = join(this.basePath, `${sanitizeSessionId(sessionId)}.jsonl`);
    if (!existsSync(sessionFile)) return null;

    try {
      const content = readFileSync(sessionFile, "utf-8");
      const firstLine = content.split("\n")[0]?.trim();
      if (firstLine) {
        const header = JSON.parse(firstLine);
        if (header.type === "header") {
          return header.previous_session_id ?? null;
        }
      }
    } catch {
      // Parse error
    }
    return null;
  }
}

// ===== Singleton =====

let _manager: SessionManager | null = null;

export function getSessionManager(): SessionManager {
  if (!_manager) {
    _manager = new SessionManager();
    logger.info({ basePath: _manager.basePath }, "SessionManager initialized");
  }
  return _manager;
}
